src/Connector/CachingConnector.php

```diff
@@ -26,23 +28,48 @@ public function __construct(CacheItemPoolInterface $cache = null)
         $this->cache = $cache ?: new MemoryCache;
     }
 
-    public function fetch($source, EncapsulatedOptions $options = null)
-    {
+    /**
+     * @param string $source
+     * @param EncapsulatedOptions|null $options
+     * @param CacheKeyGeneratorInterface $cacheKeyGenerator
+     * @return mixed
+     * @throws InvalidCacheKeyException
+     */
+    public function fetch(
+        $source,
+        EncapsulatedOptions $options = null,
+        CacheKeyGeneratorInterface $cacheKeyGenerator = null
+    ) {
         $optionsCopy = $options ? $options->copy() : [];
 
+        $key = null;
+
         if ($this->isCacheEnabled()) {
             ksort($optionsCopy);
 
-            $hash = $this->hash([$source, $optionsCopy]);
+            if ($cacheKeyGenerator !== null) {
+                $key = $cacheKeyGenerator->generateCacheKey($source, $options);
+                if (!is_string($key)) {
+                    throw new InvalidCacheKeyException('Cache key must be of type string.');
+                }
+                if (strpbrk($key, '{}()/\@:') !== false) {
+                    throw new InvalidCacheKeyException(sprintf(
+                        'Cache key "%s" contains reserved characters {}()/\@:',
+                        $key
+                    ));
+                }
+            } else {
+                $key = $this->hash([$source, $optionsCopy]);
```

Comment:
* *SourceAndOptions* does not need to be specified because every `CacheKeyGenerator` deals with *source* and *options* parameters: that is specified by the interface so it is taken for granted and does not need to comprise a portion of the class name.
* In this context, *hash* and *cache key* are the same thing so one precludes the necessity to state the other.

Considering both these points, the class name would be condensed down to just *CacheKeyGenerator*, which is the name of the interface. This indicates that, despite egregious word stuffing, not a single significant word comprises the name! Hence why I chose *JSON* as the name for this implementation.

Regarding your point about the output not actually being valid JSON, this should be a given considering PSR-6 precludes required syntax from appearing in the key, and moreover, there is no reason why it needs to be valid JSON. The name is simply used to easily identify and understand the nature of the implementation which certainly does use a JSON encoder under the hood.